### ISSUE TITLE
API formatting updates

### ISSUE THREAD

**Issue by DFurnes** [DoSomethingArchive/legacy-website#6048]
From a quick offline conversation, thinking of updating some Phoenix reportback formatting to match the changes that were [recently made in Northstar](https://github.com/DoSomething/northstar/pull/253#issuecomment-170603713).
#### Proposed Changes:
- [x] Update dates to be returned as ISO-8601 string rather than UNIX timestamps
- [x] Consider moving `pagination` into a `meta` object, and formatting `links` object for next/prev URL
- [x] Remove `retrieved_count` property from response (since getting the length of an array in PHP and JavaScript are both suuuuuper fast)
- [ ] Consider namespacing filters in query-strings, for example `/api/v1/reportback-items/?filter[status]=approved,promoted&page=2`

/cc @jonuy @aaronschachter

**Comment by jonuy** [comment]
+1 to consistent formatting even across different platforms

**Comment by chloealee** [comment]
@DFurnes @jonuy do we want dates returned as an ISO-8601 only for the API or available for the app too? Trying to decide where to put this logic (in `Transformer.php` or `Reportback.php`). cc @weerd

**Comment by DFurnes** [comment]
@chloealee I think we'd want to use a `Date` object for within the app (since that has some handy methods for working with dates built-in), and only serialize to a string when formatting into JSON.

**Comment by chloealee** [comment]
@DFurnes got it - thanks for clarifying!

**Comment by chloealee** [comment]
@weerd @DFurnes it looks like `/reportbacks` doesn't have pagination right now (but `/reportback-items` does. Should I add pagination to `/reportbacks` as well?

**Comment by weerd** [comment]
yeah! not sure why I left it off before... but it should have it :)

**Comment by chloealee** [comment]
no problemo - thanks for clarifying! On it now :)

**Comment by chloealee** [comment]
@DFurnes one more question! If we're going to update to namespacing filters in query-strings, should we be doing this for all endpoints, not just `/reportbacks`?

**Comment by DFurnes** [comment]
@chloealee Yep!

**Comment by chloealee** [comment]
@DFurnes cool! Would it make sense to separate that into it's own issue since it'll be more of a beast?

**Comment by DFurnes** [comment]
Makes sense to me. :japanese_ogre:

**Comment by chloealee** [comment]
cool I'll make a separate issue now!

**Comment by chloealee** [comment]
actually - keeping one issue but two PRs :)

**Comment by weerd** [comment]
I'm a little worried that the whole filters change approach, while useful is cutting it a little close to the launch of the mobile app. Maybe these are updates we make but hold onto for now to deploy later...? It's a rather significant change to all the URL params for all endpoints and w/out the testing coverage in place feels like a big rewrite with lots of potential for breaking coming up close to launch. Especially since @aaronschachter and @jonuy would need to change how those URL params are passed in the respective mobile apps. Just tossing this out there for discussion.

cc: @angaither

**Comment by weerd** [comment]
@DFurnes can you update the example in the description? Isn't that the current method for the filters and not what's proposed for the change with the namespacing?

```
/api/v1/reportback-items/?status=approved,promoted&page=2
```

**Comment by DFurnes** [comment]
@weerd Doh, updated the example!

Good point on the potential danger of changing the query strings so close to launch... my fear was just making sure we get the API into a "feature freeze" of where we want it to be before we're stuck maintaining compatibility with any client apps using it in the wild (and then stuck with maintaining old branching code paths or anything like that).

I'm up for holding off on changing the `filter` parameters for now though if you think it's the best choice. Worst case scenario we can probably just pay attention to app version analytics and still deprecate those within a `v1`. :+1:

**Comment by jonuy** [comment]
Thanks for flagging @weerd. Yea, it does feel iffy making a breaking change like this as we get closer to launch. But if it's a change that you feel _needs_ to happen, then I'd lean towards getting it in now. I'm _pretty_ sure on the Android side I'd just need update the query strings in [here](https://github.com/DoSomething/LetsDoThis-Android/blob/master/app/src/main/java/org/dosomething/letsdothis/network/PhoenixAPI.java). 

@aaronschachter can chime in on the iOS implications. Also, not sure if there are changes that'd need to be made on Northstar.

**Comment by DFurnes** [comment]
Northstar changes would be pretty minimal since we basically pass the unaltered request directly to Phoenix. Would just need to change the logic where we transform Northstar IDs to Phoenix IDs in signup & reportback index's `user` filter parameter.

**Comment by aaronschachter** [comment]
In theory it sounds like all we need to do on iOS is make the change [here](https://github.com/DoSomething/LetsDoThis-iOS/blob/d0ec9c176ee60a92c30ffd15920cac6af3d8b53b/Lets%20Do%20This/Controllers/Campaign/LDTCampaignViewController.m#L69) -- the only other endpoint we're using to display Reportbacks is via `signups?user=:id`.  So, doesn't sound like a big lift on the iOS side ...

**Comment by weerd** [comment]
@aaronschachter the changes to how we use query params would technically affect any and all Phoenix endpoints... not just `/reportbacks`. Not sure if that was clear or not :\

**Comment by chloealee** [comment]
Following up here - is the decision to hold on this until after the mobile app is launched? cc @DFurnes @weerd @jonuy @aaronschachter

**Comment by DFurnes** [comment]
Yeah, I think that's the safe choice since launch is so soon.

**Comment by chloealee** [comment]
cool, sounds good!
